### PR TITLE
asset conversion runtime quote api generalised

### PR DESCRIPTION
**Pull request by gilescope**
To get a quote from XYZ token to ABC token currently requires two runtime calls that then need to be combined (one call of XYZ to DOT and another DOT to ABC). This PR allows this to be done in one call and makes the quote api look more like the swap api.

It's a breaking api change, but the API has not been released on kusama yet. Devs in the field integrating this API have suggested they would like this change.

### CODE REVIEW

File: substrate/frame/asset-conversion/src/tests.rs [hunk 1]
```diff
@@ -542,6 +543,10 @@ fn can_quote_price() {
 		let user = 1;
 		let token_1 = NativeOrAssetId::Native;
 		let token_2 = NativeOrAssetId::Asset(2);
+		let asset_to_dot: BoundedVec<_, _> =
+			bvec![NativeOrAssetId::Asset(2), NativeOrAssetId::Native];
+		let dot_to_asset: BoundedVec<_, _> =
+			bvec![NativeOrAssetId::Native, NativeOrAssetId::Asset(2)];
```
Comment (joepetrowski) [inline substrate/frame/asset-conversion/src/tests.rs:549]: I don't think we should introduce "dot" to FRAME.

**Review by jsidorenko — APPROVED** [review]
LGTM

File: substrate/frame/asset-conversion/src/tests.rs [hunk 2]
```diff
@@ -604,21 +605,16 @@ fn can_quote_price() {
...
+
 		// including fee so should get less out...
 		assert_eq!(
-			AssetConversion::quote_price_exact_tokens_for_tokens(
-				NativeOrAssetId::Asset(2),
-				NativeOrAssetId::Native,
-				60,
-				true,
-			),
+			AssetConversion::quote_price_exact_tokens_for_tokens(asset_to_native.clone(), 60, true,),
```
Comment (jsidorenko) [inline substrate/frame/asset-conversion/src/tests.rs:617]:
```suggestion
			AssetConversion::quote_price_exact_tokens_for_tokens(asset_to_native.clone(), 60, true),
```

File: substrate/frame/asset-conversion/src/tests.rs [hunk 3]
```diff
@@ -627,50 +623,37 @@ fn can_quote_price() {
...
 		);
 		// including fee so should need to put more in...
 		assert_eq!(
-			AssetConversion::quote_price_tokens_for_exact_tokens(
-				NativeOrAssetId::Native,
-				NativeOrAssetId::Asset(2),
-				60,
-				true,
-			),
+			AssetConversion::quote_price_tokens_for_exact_tokens(native_to_asset.clone(), 60, true,),
```
Comment (jsidorenko) [inline substrate/frame/asset-conversion/src/tests.rs:634]:
```suggestion
			AssetConversion::quote_price_tokens_for_exact_tokens(native_to_asset.clone(), 60, true),
```

File: substrate/frame/asset-conversion/src/tests.rs [hunk 4]
```diff
@@ -627,50 +623,37 @@ fn can_quote_price() {
...
```
Comment (jsidorenko) [inline substrate/frame/asset-conversion/src/tests.rs:649]:
```suggestion
			AssetConversion::quote_price_tokens_for_exact_tokens(native_to_asset.clone(), 60, true),
```

File: substrate/frame/asset-conversion/src/tests.rs [hunk 5]
```diff
@@ -725,37 +703,87 @@ fn can_quote_price() {
...
-				amount_in,
-				false,
-			)
-			.and_then(|amount| AssetConversion::quote_price_tokens_for_exact_tokens(
-				NativeOrAssetId::Asset(2),
-				NativeOrAssetId::Native,
-				amount,
-				false,
-			)),
+			AssetConversion::quote_price_tokens_for_exact_tokens(native_to_asset, amount_in, false,)
```
Comment (jsidorenko) [inline substrate/frame/asset-conversion/src/tests.rs:718]:
```suggestion
			AssetConversion::quote_price_tokens_for_exact_tokens(native_to_asset, amount_in, false)
```

**Review by PatricioNapoli — APPROVED** [review]

File: substrate/frame/asset-conversion/src/lib.rs [hunk 6]
```diff
@@ -1018,48 +1018,44 @@ pub mod pallet {
...
+			Some(if include_fee {
+				*Self::get_amounts_out(&amount_in, &path).ok()?.last()?
 			} else {
-				None
-			}
+				let mut result = amount_in;
+				for assets_pair in path.windows(2).rev() {
+					if let [asset1, asset2] = assets_pair {
+						let (reserve_in, reserve_out) = Self::get_reserves(asset1, asset2).ok()?;
+						result = reserve_out * result / reserve_in;
```
Comment (joepetrowski) [inline substrate/frame/asset-conversion/src/lib.rs:1033]: Checked mul/div? Is it ever possible that `reserve_in` is zero?

File: cumulus/parachains/runtimes/assets/asset-hub-kusama/src/lib.rs
Comment (joepetrowski) [inline cumulus/parachains/runtimes/assets/asset-hub-kusama/src/lib.rs:1]: These changes should be made to the Fellowship repo.

File: substrate/frame/asset-conversion/src/lib.rs (hunk 6 above)
Comment (jsidorenko) [reply substrate/frame/asset-conversion/src/lib.rs:1033]: Btw we have `Self::quote()` for that, which uses the internal `mul_div()` under the hood

Comment (jsidorenko) [reply substrate/frame/asset-conversion/src/lib.rs:1033]: > Is it ever possible that reserve_in is zero?

get_reserves() will throw an error if any of the balances is zero

Comment (gilescope) [reply substrate/frame/asset-conversion/src/lib.rs]: Agreed using `Self::quote` is nicer. Have updated.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3726129

File: cumulus/parachains/runtimes/assets/asset-hub-kusama/src/lib.rs
Comment (gilescope) [reply cumulus/parachains/runtimes/assets/asset-hub-kusama/src/lib.rs:1]: Waiting on the fellowship runtimes to be updated such that kusama asset hub has the asset-conversion pallet in it. Then can address this.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3982933

File: substrate/bin/node/runtime/src/lib.rs [hunk 7]
```diff
@@ -2561,15 +2563,16 @@ impl_runtime_apis! {
...
 		Balance,
 		u128,
-		NativeOrAssetId<u32>
+		NativeOrAssetId<u32>,
+		MaxSwapLength
 	> for Runtime
 	{
-		fn quote_price_exact_tokens_for_tokens(asset1: NativeOrAssetId<u32>, asset2: NativeOrAssetId<u32>, amount: u128, include_fee: bool) -> Option<Balance> {
-			AssetConversion::quote_price_exact_tokens_for_tokens(asset1, asset2, amount, include_fee)
+		fn quote_price_exact_tokens_for_tokens(path: BoundedVec<NativeOrAssetId<u32>, MaxSwapLength>, amount: u128, include_fee: bool) -> Option<Balance> {
```
[the remark this answers is not in the document]
Comment (gilescope) [reply substrate/bin/node/runtime/src/lib.rs:2570]: We started with `Vec` originally and it got switched to `BoundedVec`.

[the remark this answers is not in the document]
Comment (bkchr) [reply substrate/bin/node/runtime/src/lib.rs:2570]: Using `BoundedVec` in a runtime api makes no real sense?

**Pull request closed without merging**